lib/srv/db/snowflake/engine.go

```diff
@@ -0,0 +1,672 @@
+// web session. Session ID replaces the session/master token returned to the Snowflake client, so only Teleport
+// has access to the Snowflake access tokens.
+func (e *Engine) processLoginResponse(bodyBytes []byte, createSessionFn func(tokens sessionTokens) (string, string, error)) ([]byte, error) {
+	loginResp, err := decodeLoginResponse(bodyBytes)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if !loginResp.Success {
+		e.Log.Errorf("Snowflake authentication failed: %s", loginResp.Message)
+		// Return not modified response, so client can handle it. Otherwise, the client my keep retrying where
+		// most likely each response will return the same error (invalid JWT when user doesn't exist for ex.)
+		return bodyBytes, nil
+	}
+
+	tokens, err := loginResp.getTokens()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sessionToken, masterToken, err := createSessionFn(tokens)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// reset tokens to prevent from endlessly growing map. Very unlikely, but you never know.
+	e.tokens.reset()
+	e.tokens.setToken(sessionToken, tokens.session.token)
+	e.tokens.setToken(masterToken, tokens.master.token)
+
+	loginResp.Data["token"] = teleportAuthHeaderPrefix + sessionToken
+	loginResp.Data["masterToken"] = teleportAuthHeaderPrefix + masterToken
+
+	newResp, err := json.Marshal(loginResp)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return newResp, err
+}
+
+func (e *Engine) handleQueryRequest(ctx context.Context, req *http.Request) (io.Reader, error) {
+	return e.modifyRequestBody(req, func(body []byte) ([]byte, error) {
+		query, err := extractSQLStmt(body)
+		if err != nil {
+			return nil, trace.Wrap(err, "failed to extract SQL query")
+		}
+
+		e.Audit.OnQuery(ctx, e.sessionCtx, common.Query{
+			Query:      query.SQLText,
+			Parameters: query.paramsToSlice(),
+		})
+
+		return body, nil
+	})
+}
+
+func (e *Engine) getConnectionToken(ctx context.Context, req *http.Request) (string, error) {
+	sessionToken := extractSnowflakeToken(req.Header)
+
+	// Authentication header always starts with "Teleport:", so we know that the token was set by us.
+	// Some SDK set this header to a "random" value (Python SDK sends None for example).
+	if !strings.Contains(sessionToken, teleportAuthHeaderPrefix) {
+		return "", nil
+	}
+
+	sessionToken = strings.TrimPrefix(sessionToken, teleportAuthHeaderPrefix)
+
+	snowflakeToken, err := e.getSnowflakeToken(ctx, sessionToken)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return snowflakeToken, nil
+}
+
+func (e *Engine) modifyRequestBody(req *http.Request, modifyReqFn func(body []byte) ([]byte, error)) (*bytes.Buffer, error) {
+	body, err := readRequestBody(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	body, err = modifyReqFn(body)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	buf := &bytes.Buffer{}
+	if req.Header.Get("Content-Encoding") == "gzip" {
```

Comment:
do we have to keep zipping (could we remove the encoding and send plain)?

Reply:
We probably could, but gzipping reduces the size of the messages. I don't see any harm in leaving it, but I do see that someone maybe unhappy with extra traffic when using Teleport.